### ISSUE TITLE
`IOSQE_BUFFER_SELECT` produces `Operation not supported` if operating on a file

### ISSUE THREAD

**Issue by ned14** [axboe/liburing#1214]
I have `IOSQE_BUFFER_SELECT` working fine on socket reads, but if I try it with file reads of an `O_DIRECT` file, I get `Operation not supported`. I am on kernel 6.8, but https://github.com/torvalds/linux/blob/master/io_uring/io_uring.c#L2032 shows trunk would have the same behaviour:

```c
		if (sqe_flags & IOSQE_BUFFER_SELECT) {
			if (!def->buffer_select)
				return io_init_fail_req(req, -EOPNOTSUPP);
```

It seems to me that the rationale for `IOSQE_BUFFER_SELECT` makes as much sense for files as for sockets in that lots of concurrent i/o on storage ties up registered i/o buffers in the same way as it does for sockets. As io_uring limits registered buffers to 4096 per ring, exhaustion of locked read i/o buffers is a real world problem for some users.

I request that `IOSQE_BUFFER_SELECT` be extended for files, at least those opened `O_DIRECT`.

**Comment by axboe** [comment]
The reason why provided buffers is only supported for sockets (and other pollable files) is that you don't necessarily know when they will become readable. This is in contrast to regular files, O_DIRECT or not makes no difference here, where the files are _always_ ready to be read. Hence there's no potentially long gap between submitting IO and it completing, only the speed of storage dictates that. Because of that, there's no gap between saying "oh I want to read from this file" and "let me pick a buffer to read from", and without that gap, it doesn't really make sense to use provided buffers - it's simpler to just pass the buffer in directly in the first place.

In other words, provided buffers do not make as much sense for storage as it does for eg sockets, as you'd save no memory using them. The buffers would be tied up exactly as long via provided buffers as they would be if you had just passed it in, in fact probably *longer* as you'd need to add a bunch to a provided buffer group first.

**Comment by ned14** [comment]
Forgive me if my knowledge of Linux kernel internals is incorrect, however isn't it the case that there are NVMe queues pinned per CPU per storage and once those get filled, the kernel keeps an aside list to be entered into the NVMe queues as they drain?

I agree that once an i/o goes into the NVMe queue, that is its buffer allocated as DMA must be done. However before that, quite a period of time can elapse (relatively speaking) if the device is under strain. If my understanding above is correct, I guess I'm asking for buffer allocation to be delayed until just before the i/o enters the NVMe queue.

~~~

Our use case is undoubtedly niche: we do a very great deal of random 512 byte reads from across the entire SSD in `O_DIRECT`. We gate the concurrency we show to io_uring to about 1000 in flight random reads, benchmarks showed that to maximise throughput for our SSD model (other models would max out at lower concurrencies). Read latencies get real high in this circumstance, and the read completion code does need to hang onto the buffer for a short while before it can release it. Hence the 4096 max locked buffer limit is uncomfortably close. I think we'll be fine for PCIe 4.0 SSDs. I'm worried about PCIe 5.0 SSDs and beyond.

Ultimately if this feature request isn't doable with the current Linux kernel structure, I guess I can live with wontfix. But that 4096 max locked buffer limit, it would be real nice if it provided more headroom soon, otherwise I would be concerned that we'll lose performance on newer SSDs due to buffer starvation. Might that be possible instead?

**Comment by axboe** [comment]
nvme generally has a submission/completion pair per CPU, unless the device has fewer queues available than the CPUs in the system. But what you describe is not how queuing works in Linux, generally things are flushed directly to the device. Hence there is no idle gap where IOs are idling waiting to get issued. Instead, if you overload the device to the extent where queuing IO runs out of resources on the device side, then that operation will fail. For io_uring, this means that we get returned a busy status, and that io_uring will then have to punt to io-wq internally to complete the operation. At that point you're way beyond the concurrency level where you can expect performance to still be improving by piling more on there, and you'd be better off simply having that IO be returned to the application with -EAGAIN as a throttling point rather than have it be a blocking attempt. I need to double check, but you should be able to do that by setting ```RWF_NOWAIT``` in the ```sqe->rw_flags``` for your storage IO. That will greatly improve efficiency for the case where you do run out of device side resources at extreme depths, and remove the point at which you're holding on to buffers because you're waiting for other IO to complete on the device.

As an aside, I'm somewhat skeptical on 1000 IOs being the sweet spot for any SSD, that's very high and way into the territory where I'd expect latencies to just be much worse at basically no increase in throughput.

For your use case, I'd strongly suggest that the key elements for the best performance may be registered buffers, not provided buffers. Those will avoid repeated page manipulations for O_DIRECT.

And finally, this one:

> Hence the 4096 max locked buffer limit

makes me a bit puzzled too, which limit is this?

**Comment by ned14** [comment]
> As an aside, I'm somewhat skeptical on 1000 IOs being the sweet spot for any SSD, that's very high and way into the territory where I'd expect latencies to just be much worse at basically no increase in throughput.

I should have been clearer: we aren't tuning for raw i/o, we're tuning for the whole system which is i/o driven, but does quite a bit of compute as well such that we run a sixteen core system at 75-95% utilisation. We have several knobs to turn, and it turns out for our very specific production hardware combination, between 500 and 1000 concurrent read i/o's before io_uring maximises performance. 

You're right that the same software on slightly lesser hardware yields a max concurrent i/o's before io_uring of around 100-200 for the whole system. If we tune for raw i/o only, we see around 64 concurrent i/o's maximises throughout without raising latency on the production hardware, and only a mere 16 concurrent i/o's on slightly lesser hardware.

Those are probably numbers more consistent with your experience. As to exactly _why_ high i/o concurrency before io_uring happens to work so well on that specific hardware, I'll be honest and say I'm not entirely sure. It could be as simple as mutexes all hitting their spincounts rather than going to sleep, or some other confounding factor. I haven't dug deep as this isn't the final software architecture, we're in the process of migrating out old hastily written layers replacing them with much higher quality layers (this is a startup). Once that final set is in place, then it'll be worth the time investment to dig into the many whys in all this and tune from there.

(Another reason I am delaying digging in is that the replacement higher quality layers have integrated ultra low overhead latency measuring designed in that is vastly superior to the current hacked in system. Yet to write the tooling to cope with the volume of data however)

> For your use case, I'd strongly suggest that the key elements for the best performance may be registered buffers, not provided buffers. Those will avoid repeated page manipulations for O_DIRECT.

We already use those.

> And finally, this one:
> 
> > Hence the 4096 max locked buffer limit
> 
> makes me a bit puzzled too, which limit is this?

There was a comment in our source code about `IORING_MAX_REG_BUFFERS` being 4096, despite that the value was set to `(1U << 14)` which is of course 16,384.

I did a quick test just there with 16,384 buffers and it passes. Which means the comment was wrong, and I came here to complain about nothing.

Sorry for the noise, and thanks for your help.
